src/team_lib/mhba/xccl_mhba_mkeys.c

```diff
@@ -0,0 +1,283 @@
+    umr_qp_creation_failed:
+    ibv_destroy_cq(ctx->umr_cq);
+    umr_cq_creation_failed:
+    return XCCL_ERR_NO_MESSAGE;
+}
+
+xccl_status_t init_umr(xccl_mhba_context_t *ctx) {
+    struct ibv_port_attr port_attr;
+    xccl_status_t status;
+    status = create_umr_qp(ctx);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    if (ibv_query_port(ctx->umr_qp->context, ctx->ib_port, &port_attr) != 0) {
+        xccl_mhba_error("Couldn't get port info");
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    status = remote_qp_connect(ctx->umr_qp, ctx->umr_qp->qp_num, port_attr.lid, ctx->ib_port);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    return XCCL_OK;
+}
+
+xccl_status_t
+create_matser_key(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node, struct mlx5dv_mkey **mkey_ptr) {
+    struct mlx5dv_mkey* mkey;
+    xccl_mhba_debug("Create MasterMKey");
+    struct mlx5dv_mkey_init_attr umr_mkey_init_attr;
+    memset(&umr_mkey_init_attr, 0, sizeof(umr_mkey_init_attr));
+    umr_mkey_init_attr.pd = ctx->ib_pd;
+    umr_mkey_init_attr.create_flags = MLX5DV_MKEY_INIT_ATTR_FLAGS_INDIRECT;
+    // Defines how many entries the Mkey will support.
+    // In case the MKey is used as "strided-KLM based MKey", the number
+    // of entries that is needed is increased by one because one entry is
+    // consumed by the "strided header" (see mlx5dv_wr_post manual).
+    umr_mkey_init_attr.max_entries = node->sbgp->group_size + 1;
+    mkey = mlx5dv_create_mkey(
+            &umr_mkey_init_attr);
+    if (mkey == NULL) {
+        xccl_mhba_debug("MasterMKey creation failed (errno=%d)", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    xccl_mhba_debug("umr_master_key_dv_mkey: lkey=0x%x", mkey->lkey);
+    *mkey_ptr = mkey;
+    return XCCL_OK;
+}
+
+xccl_status_t init_mkeys(xccl_mhba_context_t *ctx, xccl_mhba_node_t *node) {
+    xccl_status_t status;
+    status = create_matser_key(ctx, node, &node->send_mkey);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    status = create_matser_key(ctx, node, &node->recv_mkey);
+    if (status != XCCL_OK) {
+        return status;
+    }
+    xccl_mhba_debug("Creating %d KLM entries for the send mkey", node->sbgp->group_size);
+    node->send_mkey_entries = (struct mlx5dv_mr_interleaved *) calloc(node->sbgp->group_size,
+                                                                      sizeof(struct mlx5dv_mr_interleaved));
+    xccl_mhba_debug("Creating %d Strided-KLM entries for the recv mkey", node->sbgp->group_size);
+    node->recv_mkey_entries = (struct mlx5dv_mr_interleaved *) calloc(node->sbgp->group_size,
+                                                                      sizeof(struct mlx5dv_mr_interleaved));
+    return XCCL_OK;
+}
+
+xccl_status_t poll_umr_cq(xccl_mhba_context_t *ctx) {
+
+    int expected_completions_num = 1;
+    struct ibv_wc work_completion[1];
+    memset(&work_completion, 0, expected_completions_num * sizeof(struct ibv_wc));
```

Comment:
this must be wrong: &work_completion is ibv_wc **. Should be either:
&work_completion[0], OR just work_completion

Reply:
this line is from Rami's snippet. I'll check